packages/react-transition-presence/src/TransitionPresence/TransitionPresence.tsx

```diff
@@ -0,0 +1,69 @@
+import {
+  Children,
+  useCallback,
+  useEffect,
+  useMemo,
+  useState,
+  type ReactElement,
+  type ReactNode,
+} from 'react';
+import { useBeforeUnmount, type BeforeUnmountCallback } from '../useBeforeUnmount/useBeforeUnmount';
+import { TransitionPresenceContext } from './TransitionPresence.context';
+
+export type TransitionPresenceProps = {
+  children: ReactNode;
+  crossFlow?: boolean;
+};
+
+/**
+ * Will transition out old children before replacing new children. When
+ * crossFlow is enabled new children are added immediately when deferred
+ * children are not the same.
+ */
+export function TransitionPresence({ children, crossFlow }: TransitionPresenceProps): ReactElement {
+  const beforeUnmountCallbacks = useMemo(() => new Set<BeforeUnmountCallback>(), []);
+  const [deferredChildren, setDeferredChildren] = useState<ReactNode>(children);
+
+  const beforeUnmount = useCallback(
+    async (abortSignal: AbortSignal) => {
+      const promises: Array<ReturnType<BeforeUnmountCallback>> = [];
+
+      for (const callback of beforeUnmountCallbacks) {
+        promises.push(callback(abortSignal));
+      }
```

Comment:
Why no `const promises = beforeUnmountCallbacks.map(callback => callback(abortSignal))`? :)

Or do you expect there could be more things added to that list, and you want to have things clear and separated?

Reply:
The `beforeUnmountCallbacks` is a `Set`, there's no .map function on the iterator of a `Set`.

Reply:
Ah yes, and using `Set.values` is a bit too much here :) 👍